crates/sui-adapter/src/programmable_transactions/execution.rs

```diff
@@ -481,6 +578,54 @@ fn vm_move_call<E: fmt::Debug, S: StorageView<E>>(
     Ok(result)
 }
 
+fn upgrade_check_compatiblity_and_publish_package<E: fmt::Debug, S: StorageView<E>>(
+    context: &ExecutionContext<E, S>,
+    current_package: &MovePackage,
+    upgraded_package_modules: Vec<CompiledModule>,
+) -> Result<Object, ExecutionError> {
+    // Check compatibility
+    let check_struct_and_pub_function_linking = true;
+    let check_struct_layout = true;
+    let check_friend_linking = false;
+    let compatibility = Compatibility::new(
+        check_struct_and_pub_function_linking,
+        check_struct_layout,
+        check_friend_linking,
+    );
+
+    // The current package should be able to be deserialized and normalized otherwise something has gone terribly wrong.
+    let Ok(current_normalized) = current_package.normalize() else {
+            invariant_violation!("Tried to normalize modules in already-published package but failed")
+    };
+
+    let mut new_normalized = normalize_deserialized_modules(upgraded_package_modules.iter());
+
+    for (name, cur_module) in current_normalized {
+        let Some(new_module) = new_normalized.remove(&name) else {
+            return Err(ExecutionError::from_kind(ExecutionErrorKind::PackageUpgradeError(PackageUpgradeError::IncompatibleUpgrade)))
```

Comment:
nit: Is there a way for us to explain the nature of this error, like we do below for the individual compatibility issues?

Reply:
I'm pretty sure we can. `ExistingModuleNotInUpgradedPackage(Identifier)` or something like that with a corresponding explanation of "Existing module {name} in package not found in next version of package" (or something like that).